File: backend/app/schemas/ai.py
"""
Pydantic schemas for AI Platform v0.4: Copilot, Knowledge, Memory, Skills, MCP, Learning, Evals.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, StrictBool, model_validator

# --- Copilot & Suggestions ---


class AISuggestionDTO(BaseModel):
    id: int
    conversation_id: int
    inbound_message_id: int | None = None
    ai_run_id: int | None = None
    suggested_text: str
    status: str  # pending | accepted | edited | rejected | expired | auto_sent
    generated_at: datetime
    reviewed_at: datetime | None = None
    reviewed_by: str | None = None
    final_message_id: int | None = None
    edit_distance: int | None = None
    edit_ratio: float | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class AcceptSuggestionRequest(BaseModel):
    suggestion_id: int | None = None


class EditSuggestionRequest(BaseModel):
    edited_text: str = Field(min_length=1, max_length=10000)
    suggestion_id: int | None = None


class RejectSuggestionRequest(BaseModel):
    reason: str | None = None
    suggestion_id: int | None = None


# --- AI Runs & Traces ---


class AIRunDTO(BaseModel):
    id: int
    trace_id: str
    conversation_id: int
    input_message_id: int | None = None
    model: str | None = None
    provider: str | None = None
    prompt_version: str | None = None
    skills: list[str] = Field(default_factory=list)
    retrieval: list[dict[str, Any]] = Field(default_factory=list)
    citations: list[dict[str, Any]] = Field(default_factory=list)
    memory: list[dict[str, Any]] = Field(default_factory=list)
    memories: list[dict[str, Any]] = Field(default_factory=list)
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    rag_hit_count: int = 0
    tool_call_count: int = 0
    decision: str
    confidence: float | None = None
    latency_ms: int | None = None
    tokens: int | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    cached_input_tokens: int | None = None
    reasoning_tokens: int | None = None
    llm_call_count: int | None = None
    usage_source: str = "unavailable"
    estimated_cost: float | None = None
    cost_currency: str = "USD"
    traffic_source: str = "production"
    errors: str | None = None
    final_message_id: int | None = None
    created_at: datetime
    model_calls: list[dict[str, Any]] = Field(default_factory=list)
    attachments: list[dict[str, Any]] = Field(default_factory=list)


class AIRunListResponse(BaseModel):
    items: list[AIRunDTO]
    total: int
    limit: int
    offset: int


# --- Knowledge Base & RAG ---


class KnowledgeSourceDTO(BaseModel):
    id: int
    title: str
    source_type: str  # manual | faq | document | product | conversation | website | api
    source_uri: str | None = None
    language: str
    status: str
    trust_level: str
    version: int
    created_by: str
    created_at: datetime
    updated_at: datetime
    documents_count: int = 0


class KnowledgeDocumentDTO(BaseModel):
    id: int
    source_id: int
    title: str
    content: str
    scope_type: str
    scope_id: str | None = None
    chunk_count: int = 0
    created_at: datetime
    updated_at: datetime


class KnowledgeSourceDetailDTO(BaseModel):
    source: KnowledgeSourceDTO
    documents: list[KnowledgeDocumentDTO] = Field(default_factory=list)


class CreateKnowledgeSourceRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    source_type: str = Field(
        default="manual", pattern="^(manual|faq|document|product|conversation|website|api)$"
    )
    source_uri: str | None = None
    language: str = "en"
    trust_level: str = "standard"


class CreateDocumentRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1)
    scope_type: str = Field(default="global", pattern="^(global|group|user)$")
    scope_id: str | None = None


class CreateFAQRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    answer: str = Field(min_length=1, max_length=5000)
    category: str = "general"
    scope_type: str = "global"
    scope_id: str | None = None


class KnowledgeSearchRequest(BaseModel):
    query: str = Field(min_length=1)
    limit: int = Field(default=5, ge=1, le=20)
    group_id: str | None = None
    user_id: str | None = None


class KnowledgeSearchResultDTO(BaseModel):
    chunk_id: int
    document_id: int
    source_id: int
    title: str
    content: str
    score: float
    scope_type: str
    scope_id: str | None = None


# --- Memory ---


class MemoryItemDTO(BaseModel):
    id: int
    scope_type: str  # user | group | global
    scope_id: str
    memory_type: str  # preference | fact | case | constraint
    content: str
    confidence: float = 1.0
    importance: int = 3
    sensitivity: str = "standard"
    status: str = "active"
    created_by: str = "ai_extraction"
    created_at: datetime
    updated_at: datetime | None = None
    is_pii_redacted: bool = False


class CreateMemoryRequest(BaseModel):
    scope_type: str = Field(default="user", pattern="^(user|group|global)$")
    scope_id: str = Field(min_length=1)
    memory_type: str = Field(default="fact", pattern="^(preference|fact|case|constraint)$")
    content: str = Field(min_length=1, max_length=2000)
    importance: int = Field(default=3, ge=1, le=5)


# --- Skills ---


class SkillDTO(BaseModel):
    name: str
    description: str
    version: str
    scope: str
    is_enabled: bool
    body: str | None = None


class ToggleSkillRequest(BaseModel):
    is_enabled: bool


# --- MCP ---


class MCPServerDTO(BaseModel):
    id: int
    name: str
    transport: str  # stdio | http | sse
    command_or_url: str
    is_enabled: bool
    status: str
    last_connected_at: datetime | None = None
    error_message: str | None = None
    tools_count: int = 0
    # Backward compatibility aliases
    transport_type: str | None = None
    command: str | None = None
    endpoint_url: str | None = None
    tool_count: int | None = None
    last_health_check: datetime | None = None


class MCPServerDetailDTO(BaseModel):
    server: MCPServerDTO
    discovered_tools: list[dict[str, Any]] = Field(default_factory=list)
    # Convenience flattened fields for UI compatibility
    id: int | None = None
    name: str | None = None
    transport: str | None = None
    transport_type: str | None = None
    endpoint_url: str | None = None
    command: str | None = None
    is_enabled: bool | None = None
    status: str | None = None
    tools: list[dict[str, Any]] = Field(default_factory=list)


class CreateMCPServerRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    transport: str = Field(default="stdio")
    transport_type: str | None = None
    command_or_url: str = Field(default="", max_length=512)
    command: str | None = None
    endpoint_url: str | None = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)

    @model_validator(mode="before")
    @classmethod
    def normalize_fields(cls, data: Any) -> Any:
        if not isinstance(data, dict):
            return data
        data = dict(data)
        raw = data.get("transport") or data.get("transport_type") or "stdio"
        if not isinstance(raw, str):
            raise ValueError("transport must be a string")
        transport = raw.strip().lower()
        if transport not in ("stdio", "http", "sse"):
            raise ValueError(f"Unsupported transport '{raw}'. Expected stdio, http, or sse.")
        data["transport"] = transport
        if not data.get("command_or_url"):
            if transport == "stdio":
                data["command_or_url"] = data.get("command") or ""
            else:
                data["command_or_url"] = data.get("endpoint_url") or data.get("command") or ""
        if not str(data.get("command_or_url") or "").strip():
            raise ValueError("command_or_url (or command/endpoint_url) is required")
        return data


class UpdateMCPServerRequest(BaseModel):
    is_enabled: StrictBool | None = None


# --- Learning Loop ---


class LearningCandidateDTO(BaseModel):
    id: int
    conversation_id: int
    customer_question: str
    human_answer: str
    suggested_faq_q: str | None = None
    suggested_faq_a: str | None = None
    category: str
    language: str
    source_quality: str
    status: str  # pending | approved | rejected | promoted
    reviewed_by: str | None = None
    reviewed_at: datetime | None = None
    created_at: datetime


class PromoteCandidateRequest(BaseModel):
    action: Literal["knowledge", "training"] = "knowledge"
    faq_question: str | None = None
    faq_answer: str | None = None
    category: str = "general"
    scope_type: str = Field(default="global", pattern="^(global|group|user)$")
    scope_id: str | None = None
    confirm_global_privacy: bool = False


class TrainingStatsDTO(BaseModel):
    total_approved_examples: int


# --- Prompts ---


class PromptVersionDTO(BaseModel):
    id: int
    version: str
    name: str
    template: str
    is_active: bool
    created_by: str
    created_at: datetime
    activated_at: datetime | None = None


class CreatePromptVersionRequest(BaseModel):
    version: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    template: str = Field(min_length=10, max_length=50000)
    set_active: bool = False


# --- Evaluation & Overview ---


class EvaluationSuiteResultDTO(BaseModel):
    total_cases: int
    passed_cases: int
    pass_rate: float
    decision_accuracy: float
    average_latency_ms: float
    total_tokens: int | None = None
    results: list[dict[str, Any]]
    run_id: int | None = None
    eval_type: str = "deterministic"


class EvaluationRunDTO(BaseModel):
    id: int
    eval_type: str
    provider: str | None = None
    model: str | None = None
    prompt_version: str | None = None
    dataset_version: str | None = None
    total_cases: int
    passed_cases: int
    pass_rate: float
    decision_accuracy: float
    average_latency_ms: float
    total_tokens: int | None = None
    estimated_cost: float | None = None
    status: str
    started_at: datetime
    completed_at: datetime | None = None
    results: list[dict[str, Any]] = Field(default_factory=list)


class ProviderLiveTestRequest(BaseModel):
    test_tools: bool = True
    test_embeddings: bool = True


class ProviderLiveTestResponse(BaseModel):
    connected: bool
    connection_status: str = "connected"
    provider: str
    provider_detected: str = "openai_compatible"
    model: str
    endpoint: str
    latency_ms: int = 0
    preview: str = ""
    response_preview: str = ""
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    cached_tokens: int | None = None
    reasoning_tokens: int | None = None
    usage_source: str = "unavailable"
    finish_reason: str | None = None
    tool_calling_status: str = "not_verified"
    native_tool_calling: str = "not_verified"
    embedding_status: str = "not_configured"
    embeddings: str = "not_configured"
    streaming_status: str = "not_verified"
    vision_status: str = "not_verified"
    audio_status: str = "not_verified"
    tested_at: datetime
    error: str | None = None
    error_message: str | None = None


# --- Usage Telemetry ---


class UsageSummaryDTO(BaseModel):
    time_range: str
    total_runs: int = 0
    total_tokens: int = 0
    input_tokens: int | None = None
    output_tokens: int | None = None
    cached_input_tokens: int | None = None
    reasoning_tokens: int | None = None
    estimated_cost: float | None = None
    currency: str = "USD"
    cost_currency: str = "USD"
    total_model_calls: int = 0
    avg_latency_ms: float = 0.0
    errors: int = 0
    error_rate: float = 0.0


class UsageTimeseriesPointDTO(BaseModel):
    timestamp: str
    runs: int = 0
    errors: int = 0
    total_tokens: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    avg_latency_ms: float = 0.0


class UsageTimeseriesResponseDTO(BaseModel):
    points: list[UsageTimeseriesPointDTO]
    time_range: str
    timezone: str = "UTC"


class ModelUsageItemDTO(BaseModel):
    provider: str
    model: str
    runs: int
    errors: int
    error_rate: float
    input_tokens: int
    output_tokens: int
    total_tokens: int
    avg_latency_ms: float
    estimated_cost: float | None = None
    currency: str = "USD"


class ModelUsageResponseDTO(BaseModel):
    models: list[ModelUsageItemDTO]


class AIOverviewMetricsDTO(BaseModel):
    total_ai_runs: int = 0
    total_runs: int = 0
    successful_runs: int = 0
    error_runs: int = 0
    error_rate: float = 0.0
    automation_rate: float = 0.0
    copilot_rate: float = 0.0
    human_takeover_rate: float = 0.0
    suggestion_acceptance_rate: float = 0.0
    copilot_acceptance_rate: float = 0.0
    copilot_suggestions_count: int = 0
    suggestion_edit_rate: float = 0.0
    copilot_avg_edit_ratio: float = 0.0
    suggestion_rejection_rate: float = 0.0
    rag_hit_rate: float = 0.0
    tool_call_success_rate: float = 0.0
    tool_approval_rate: float = 0.0
    memory_items_count: int = 0
    knowledge_documents_count: int = 0
    knowledge_sources_count: int = 0
    knowledge_chunks_count: int = 0
    average_latency_ms: float = 0.0
    avg_latency_ms: float = 0.0
    p50_latency_ms: float | None = None
    p95_latency_ms: float | None = None
    p99_latency_ms: float | None = None
    total_tokens_used: int = 0
    total_tokens: int = 0
    input_tokens: int | None = None
    output_tokens: int | None = None
    cached_input_tokens: int | None = None
    reasoning_tokens: int | None = None
    tokens_source: str = "unavailable"
    estimated_cost: float | None = None
    cost_currency: str = "USD"
    time_range: str = "all"
    timezone: str = "UTC"

File: frontend/src/types/ai.ts
/**
 * AI Platform v0.4 Domain Types.
 */

export interface AISuggestionDTO {
  id: number;
  conversation_id: number;
  inbound_message_id?: number | null;
  ai_run_id?: number | null;
  suggested_text: string;
  status: 'pending' | 'accepted' | 'edited' | 'rejected' | 'expired' | 'auto_sent';
  generated_at: string;
  reviewed_at?: string | null;
  reviewed_by?: string | null;
  edit_distance?: number | null;
  edit_ratio?: number | null;
  metadata?: {
    citations?: Array<{
      chunk_id?: number;
      title?: string;
      snippet?: string;
      score?: number;
    }>;
    skills_used?: string[];
    memories_used?: Array<{
      id?: number;
      content?: string;
      type?: string;
    }>;
    tools_called?: Array<{
      name?: string;
      arguments?: Record<string, unknown>;
    }>;
  } | null;
}

export interface AIRunExplainabilityDTO {
  id: number;
  trace_id: string;
  conversation_id: number;
  model?: string | null;
  provider?: string | null;
  prompt_version?: string | null;
  decision: string;
  confidence?: number | null;
  latency_ms?: number | null;
  tokens?: number | null;
  skills: string[];
  citations: Array<{
    chunk_id?: number;
    title?: string;
    snippet?: string;
    score?: number;
  }>;
  memories: Array<{
    id?: number;
    content?: string;
    type?: string;
  }>;
  tool_calls: Array<{
    name?: string;
    arguments?: Record<string, unknown>;
  }>;
  attachments?: Array<{
    id?: number;
    filename?: string | null;
    mime_type?: string | null;
    processing_status?: string | null;
    extracted_text?: string | null;
    processor_model?: string | null;
    processor_type?: string | null;
  }>;
  created_at: string;
}

export interface AIOverviewMetricsDTO {
  total_ai_runs?: number;
  total_runs: number;
  automation_rate?: number;
  copilot_rate?: number;
  human_takeover_rate?: number;
  suggestion_acceptance_rate?: number;
  copilot_acceptance_rate: number;
  copilot_suggestions_count?: number;
  suggestion_edit_rate?: number;
  copilot_avg_edit_ratio?: number;
  suggestion_rejection_rate?: number;
  rag_hit_rate?: number;
  knowledge_documents_count?: number;
  knowledge_sources_count?: number;
  knowledge_chunks_count?: number;
  memory_items_count: number;
  average_latency_ms?: number;
  avg_latency_ms: number;
  p50_latency_ms?: number | null;
  p95_latency_ms?: number | null;
  p99_latency_ms?: number | null;
  total_tokens_used?: number;
  total_tokens: number;
  input_tokens?: number | null;
  output_tokens?: number | null;
  cached_input_tokens?: number | null;
  reasoning_tokens?: number | null;
  estimated_cost?: number | null;
  cost_currency?: string;
  successful_runs?: number;
  error_runs?: number;
  error_rate?: number;
  tool_invocation_count?: number;
  tool_call_success_rate?: number;
  tool_approval_rate?: number;
  active_prompt_version?: string | null;
  time_range?: string;
  timezone?: string;
}

export interface AIModelCallDTO {
  id: number;
  phase: string;
  provider?: string | null;
  model?: string | null;
  latency_ms: number;
  input_tokens?: number | null;
  output_tokens?: number | null;
  total_tokens?: number | null;
  cached_input_tokens?: number | null;
  reasoning_tokens?: number | null;
  usage_source: string;
  finish_reason?: string | null;
  provider_request_id?: string | null;
  success: boolean;
  error?: string | null;
  estimated_cost?: number | null;
  currency?: string;
  started_at?: string;
}

export interface AIRunDTO {
  id: number;
  trace_id: string;
  conversation_id: number;
  input_message_id?: number | null;
  final_message_id?: number | null;
  model?: string | null;
  provider?: string | null;
  prompt_version?: string | null;
  decision: string;
  confidence?: number | null;
  latency_ms?: number | null;
  tokens?: number | null;
  input_tokens?: number | null;
  output_tokens?: number | null;
  total_tokens?: number | null;
  cached_input_tokens?: number | null;
  reasoning_tokens?: number | null;
  llm_call_count?: number;
  usage_source?: string;
  estimated_cost?: number | null;
  cost_currency?: string;
  traffic_source?: string;
  rag_hit_count?: number;
  tool_call_count?: number;
  retrieval?: Array<Record<string, unknown>>;
  citations: Array<{
    chunk_id?: number;
    title?: string;
    snippet?: string;
    score?: number;
  }>;
  memory?: Array<Record<string, unknown>>;
  memories: Array<{
    id?: number;
    content?: string;
    type?: string;
  }>;
  tool_calls: Array<{
    name?: string;
    arguments?: Record<string, unknown>;
    status?: string;
  }>;
  model_calls?: AIModelCallDTO[];
  errors?: string | null;
  created_at: string;
}

export interface AIRunListResponse {
  items: AIRunDTO[];
  total: number;
  limit: number;
  offset: number;
}

export interface UsageSummaryDTO {
  time_range?: string;
  total_runs: number;
  total_tokens: number;
  input_tokens: number | null;
  output_tokens: number | null;
  cached_input_tokens: number | null;
  reasoning_tokens: number | null;
  estimated_cost: number | null;
  currency: string;
  cost_currency: string;
  total_model_calls: number;
  avg_latency_ms: number;
  errors: number;
  error_rate: number;
}

export interface UsageTimeseriesPointDTO {
  timestamp: string;
  runs: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  errors: number;
  avg_latency_ms: number;
}

export interface UsageTimeseriesResponseDTO {
  points: UsageTimeseriesPointDTO[];
  time_range: string;
  timezone: string;
}

export interface ModelUsageItemDTO {
  provider: string;
  model: string;
  runs: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  cached_input_tokens?: number;
  reasoning_tokens?: number;
  avg_latency_ms: number;
  errors?: number;
  error_count?: number;
  error_rate?: number;
  estimated_cost: number | null;
  currency: string;
}

export interface ModelUsageResponseDTO {
  models: ModelUsageItemDTO[];
}

export interface KnowledgeDocumentDTO {
  id: number;
  source_id: number;
  title: string;
  content: string;
  scope_type: string;
  scope_id?: string | null;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeSourceDetailDTO {
  source: KnowledgeSourceDTO;
  documents: KnowledgeDocumentDTO[];
}

export interface RAGSearchResultDTO {
  chunk_id: number;
  document_id: number;
  source_id: number;
  title: string;
  snippet: string;
  score: number;
  scope_type: string;
  scope_id?: string | null;
}

export interface MCPToolDetailDTO {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
  read_only: boolean;
  requires_approval: boolean;
  has_external_effects: boolean;
}

export interface MCPServerDetailDTO extends MCPServerDTO {
  tools: MCPToolDetailDTO[];
  discovered_tools?: MCPToolDetailDTO[];
  server?: MCPServerDTO;
  latency_ms?: number | null;
}

export interface ProviderLiveTestResultDTO {
  endpoint: string;
  model: string;
  provider_detected: string;
  connection_status: 'connected' | 'failed' | 'not_configured' | string;
  connected?: boolean;
  latency_ms?: number | null;
  preview?: string | null;
  response_preview?: string | null;
  input_tokens?: number | null;
  output_tokens?: number | null;
  total_tokens?: number | null;
  usage_source?: string;
  native_tool_calling: 'supported' | 'unsupported' | 'not_verified' | 'error' | string;
  tool_calling_status?: string;
  embeddings: 'connected' | 'not_configured' | 'unsupported' | 'error' | string;
  embedding_status?: string;
  tested_at: string;
  error?: string | null;
  error_message?: string | null;
}

export interface EvaluationRunDTO {
  id: number;
  eval_type: string;
  status: string;
  provider?: string | null;
  model?: string | null;
  prompt_version?: string | null;
  dataset_version?: string | null;
  total_cases: number;
  passed_cases: number;
  pass_rate: number;
  decision_accuracy: number;
  average_latency_ms: number;
  total_tokens: number | null;
  estimated_cost?: number | null;
  started_at: string;
  completed_at?: string | null;
  results: EvaluationResultCaseDTO[] | Record<string, unknown>[];
}

export interface TrainingStatsDTO {
  total_approved_examples: number;
  languages?: Record<string, number>;
}

export interface KnowledgeSourceDTO {
  id: number;
  title: string;
  source_type: string;
  source_uri?: string | null;
  language: string;
  status: string;
  trust_level: string;
  version: number;
  document_count?: number;
  documents_count?: number;
  chunk_count?: number;
  chunks_count?: number;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface MemoryItemDTO {
  id: number;
  scope_type: 'user' | 'group' | 'global';
  scope_id: string;
  content: string;
  memory_type: string;
  confidence?: number;
  importance: number;
  sensitivity?: string;
  status?: string;
  is_pii_redacted?: boolean;
  created_by: string;
  created_at: string;
  updated_at?: string | null;
}

export interface SkillDTO {
  name: string;
  description: string;
  version: string;
  scope: string;
  is_enabled: boolean;
  instructions_preview?: string;
  body?: string | null;
}

export interface MCPServerDTO {
  id: number;
  name: string;
  transport?: string;
  transport_type?: string;
  command_or_url?: string;
  endpoint_url?: string | null;
  command?: string | null;
  status: string;
  is_enabled: boolean;
  tools_count?: number;
  tool_count?: number;
  last_connected_at?: string | null;
  last_health_check?: string | null;
  error_message?: string | null;
}

export interface LearningCandidateDTO {
  id: number;
  conversation_id: number;
  customer_question: string;
  human_answer: string;
  suggested_faq_q?: string | null;
  suggested_faq_a?: string | null;
  category: string;
  language: string;
  source_quality: string;
  status: 'pending' | 'promoted' | 'approved' | 'rejected';
  created_at: string;
}

export interface EvaluationResultCaseDTO {
  case_id: string;
  case_name: string;
  passed: boolean;
  decision_correct: boolean;
  actual_decision: string;
  keyword_score: number;
  latency_ms: number;
  tokens: number;
}

export interface EvaluationSummaryDTO {
  total_cases: number;
  passed_cases: number;
  pass_rate: number;
  decision_accuracy: number;
  average_latency_ms: number;
  total_tokens: number | null;
  results: EvaluationResultCaseDTO[];
}

export interface PromptVersionDTO {
  id: number;
  version: string;
  name: string;
  template: string;
  is_active: boolean;
  created_by: string;
  created_at: string;
  activated_at?: string | null;
}

export interface AIDiagnosticsDTO {
  llm: { name?: string; model?: string; status: string };
  llm_provider?: { name?: string; model?: string; status: string };
  embedding: { name?: string; model?: string; status: string };
  embedding_provider?: { name?: string; model?: string; status: string };
  vector_store: { provider: string; status: string };
  rag_index: { documents_count: number; chunks_count: number; status: string };
  mcp: { active_servers: number; total_servers: number; status: string };
  signal_gateway: { status: string; api_url?: string; phone_number?: string };
  realtime?: {
    active_subscribers: number;
    total_events_emitted: number;
    dropped_events: number;
    status?: string;
  };
  capabilities?: {
    text?: string;
    streaming?: string;
    vision?: string;
    audio?: string;
    multimodal?: string;
  };
}



